modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/transaction/TransactionStorage.scala

```diff
@@ -96,51 +114,100 @@ class TransactionStorage[F[_]: Async: KryoSerializer](
+            val lastTxState = lastAccepted.getOrElse(address, Majority(TransactionReference.empty))
+            Consecutive
+              .take[F](waiting.toList, lastTxState.ref)
+              .map(consecutiveTxs => pullForAddress(lastTxState, consecutiveTxs))
           case None =>
             List.empty[Hashed[Transaction]].pure[F]
         }
       }.map(_.flatten)
     } yield txs.size
 
-  def pull(): F[Option[NonEmptyList[Hashed[Transaction]]]] =
+  def pull(count: NonNegLong): F[Option[NonEmptyList[Hashed[Transaction]]]] =
     for {
       lastAccepted <- lastAccepted.toMap
-      addresses <- waitingTransactions.keys
+      addresses <- waitingTransactions.keys.map(_.sorted)
       allPulled <- addresses.traverse { address =>
         val pulledM = for {
           (maybeWaiting, setter) <- waitingTransactions(address).access
 
-          maybePulled <- maybeWaiting.traverse { waiting =>
-            val lastTx = lastAccepted.getOrElse(address, TransactionReference.empty)
+          pulled <- maybeWaiting.traverse { waiting =>
+            val lastTxState = lastAccepted.getOrElse(address, Majority(TransactionReference.empty))
             Consecutive
-              .take[F](waiting.toList, lastTx)
-              .map { pulled =>
-                (SortedSet.from(waiting.toList.diff(pulled)).toNes, pulled)
-              }
+              .take[F](waiting.toList, lastTxState.ref)
+              .map(consecutiveTxs => pullForAddress(lastTxState, consecutiveTxs))
+              .map(pulled => (SortedSet.from(waiting.toList.diff(pulled)).toNes, pulled))
               .flatMap {
                 case (maybeNotPulled, pulled) =>
                   val maybeStillWaiting = maybeNotPulled
-                    .flatMap(_.filter(_.ordinal > lastTx.ordinal).toNes)
+                    .flatMap(_.filter(_.ordinal > lastTxState.ref.ordinal).toNes)
 
                   setter(maybeStillWaiting)
                     .ifM(
-                      logger.debug(s"Pulled ${pulled.size} transaction(s) for consensus").as(pulled),
+                      NonEmptyList.fromList(pulled).pure[F],
                       logger.debug("Concurrent update occurred while trying to pull transactions") >>
-                        List.empty[Hashed[Transaction]].pure[F]
+                        none[NonEmptyList[Hashed[Transaction]]].pure[F]
                     )
               }
-          }
-          pulled = maybePulled.getOrElse(List.empty)
+          }.map(_.flatten)
         } yield pulled
 
         pulledM.handleErrorWith {
           logger.warn(_)(s"Error while pulling transactions for address=${address.show} from waiting pool.") >>
-            List.empty.pure[F]
+            none[NonEmptyList[Hashed[Transaction]]].pure[F]
         }
       }.map(_.flatten)
-    } yield NonEmptyList.fromList(allPulled)
+
+      selected = takeFirstNHighestFeeTxs(allPulled, count)
+      toReturn = allPulled.flatMap(_.toList).toSet.diff(selected.toSet)
```

Comment:
instead of pulling, checking and returning, we can just implement the limit in consecutive and do not pull more than expected

Reply:
We don't how many of these consecutive transactions for a single address will be pulled, because we first form the chains and only then take transactions by highest fee from all chains for all addresses. We don't know if another addresses will or won't have more transactions with higher fee and hence our transactions will have to wait. I don't know how many should I pull at the moment I'm calling consecutive.

Reply:
that's true but imagine having a lot of transactions in a queue (100k, 500k). you'd shuffle them all back and forth. maybe we should keep them in consecutive order so we could see how many txs there are without moving them?

Reply:
The problem is the fees and the not amount of the transactions. We need to give the precedence to the transactions with highest fee.

Reply:
I could also change the `MapRef[Address, List[Transaction]]` to `Ref[Map[Address, List[Transaction]]]` and do all these operations in a `.modify` but then uploading the transactions may be less performant as I will block the whole ref instead of just one address in a mapref.